### PR TITLE
WorkerTaskMonitor: stop() waits for mainLoop to exit.

### LINKED ISSUE #2637: Transient failure in WorkerTaskMonitorTest.testRestartCleansOldStatus
https://travis-ci.org/druid-io/druid/builds/115348852

```
Tests run: 4, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.356 sec <<< FAILURE! - in io.druid.indexing.worker.WorkerTaskMonitorTest
testRestartCleansOldStatus(io.druid.indexing.worker.WorkerTaskMonitorTest)  Time elapsed: 1.213 sec  <<< FAILURE!
java.lang.AssertionError: expected:<FAILED> but was:<RUNNING>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at io.druid.indexing.worker.WorkerTaskMonitorTest.testRestartCleansOldStatus(WorkerTaskMonitorTest.java:338)
```

### PR DESCRIPTION
**Pull request by gianm**
Fixes #2637.

### CODE REVIEW

File: indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java [hunk 1]
```diff
@@ -237,18 +242,21 @@ public void onFailure(Throwable t)
...
       try {
+        log.info("Stopping WorkerTaskMonitor.");
+        started = false;
         exec.shutdownNow();
         pathChildrenCache.close();
         taskRunner.stop();
-
-        started = false;
-        log.info("Stopped WorkerTaskMonitor.");
+        doneStopping.await();
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java]: How does this timeout?

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java]: It doesn't, do you think it should? should the value be configurable?

My attitude is generally that program-lifecycle stop() things don't need to time out, as people should have some higher level thing that will kill -9 the proc if it takes too long to shut down. But we can have stop() timeouts if we want.

Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java]: I think there should be a way for someone to be able to break out of a broken latch. I thiiiiink the user should be able to send another -15 to get this to interrupt? 

If so then I think a simple log before this await indicating that you are about to wait for shutdown to complete should suffice.

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java]: sure, I will add a log message.

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java]: updated.

**Comment by drcrallen** [comment]
Suggest adding sanity check in start to ensure latch hasn't already fired, like if getCount is not 1 throw ISE on start.

**Comment by drcrallen** [comment]
that way noone in the future can start this twice.

**Comment by gianm** [comment]
@drcrallen I think the sanity checks already there will prevent starting twice. (ISE gets thrown if "started" or "exec.isShutdown" is true)

**Comment by drcrallen** [comment]
ah, good call, :+1:  then

**Comment by fjy** [comment]
:+1:

**Pull request merged**